### PR TITLE
Feature/tools refactor

### PR DESCRIPTION
**Pull request by michaelfuest**
## 📦 Pull Request Summary

This PR introduces a major refactoring of the workflow engine to use a tool-based architecture, enabling more autonomous, reliable and maintainable workflow building and execution through structured LLM tool calls.

### ✨ What's Changed

- **Workflow Engine**:
  - Introduced new tool-based workflow building system (`workflow-tools.ts`)
  - Refactored `WorkflowBuilder` to use structured tool calls instead of direct LLM generation
  - Refactored execution self-healing to execute in an agentic loop
  - Removed and deprecated `dataPath` and `inputMapping` from new generated workflows while maintaining backwards compatibility
  - Removed JSONata expressions from all newly generated workflows, they will only exist in legacy workflows moving forward

- **LLM Infrastructure**:
  - Added `executeTaskWithTools` pattern to all LLM providers (OpenAI, Anthropic, Gemini)
  - Implemented structured tool calling with proper error handling and retry logic
  - Added tool execution tracing and abort conditions
  - Added tool execution, calling and result interfaces

- **API & Self-Healing**:
  - Now using executable stopConditions to determine when to stop paginating through results
  - Ensured backwards compatibility for workflows w/o stop conditions

- **Other**:
  - Updated documentation crawling to be more extensive
  - Refining documentation keyword searching logic
---

### 🎯 Motivation / Context

- The main motivation for this PR is to improve reliability of the core workflow execution engine by giving the agent more autonomy during execution (or even introducing an agentic pattern in the first place). Another motivation was to improve the workflow / LLM architecture to be more extensible. Since (almost) all LLM calling logic now involves structured tool calls, it is very easy to provide additional capabilities to agents (like web search) to agents where we see a need. 
---

## ✅ Contributor Checklist

- [ ] Documentation is up-to-date
- [ ] MCP tool descriptions and instructions are up-to-date
- [x] Client SDK updated (if applicable)
- [ ] Integration test suite has been run and passes (may not be necessary)
- [ ] Unit tests added or updated (if applicable)
- [ ] Changes are covered by tests

---

## ⚠️ Compatibility Notes

- [ ] ✅ Fully backward compatible
- [x] ⚠️ Minor behavioral change (non-breaking)
- [ ] ❌ Breaking change — migration steps documented below

<details>
<summary>🔁 Migration Notes (if applicable)</summary>

No migration required. All changes maintain backwards compatibility:
- Existing workflows with `dataPath` will continue to work
- Old pagination logic without `stopCondition` is preserved
- Old input and response mappings will continue to work
- JSONata mappings are still executable for legacy workflows

</details>

---

## 📝 Additional Notes

- The new tool-based architecture sets the foundation for future enhancements like more sophisticated error recovery, web search and re-building from scratch.

### CODE REVIEW

File: packages/core/eval/integration-testing/integration-test-config.json [hunk 1]
```diff
@@ -72,8 +72,8 @@
             "jira": {
                 "id": "jira",
                 "name": "JIRA",
-                "urlHost": "https://superglue-team-test.atlassian.net",
-                "urlPath": "/rest/api/3",
+                "urlHost": "https://superglue-team-test.atlassian.net/",
```
Comment (stefanfaistenauer) [inline packages/core/eval/integration-testing/integration-test-config.json]: not sure this does anything

File: packages/core/eval/utils/setup-manager.ts [hunk 2]
```diff
@@ -287,7 +287,7 @@ export class SetupManager {
             await waitForIntegrationProcessing(
                 datastoreAdapter,
                 pendingIntegrations,
-                240000
+                600000
```
Comment (stefanfaistenauer) [inline packages/core/eval/utils/setup-manager.ts]: i assume timeout? good.

File: packages/core/graphql/resolvers/call.ts [hunk 3]
```diff
@@ -23,86 +25,173 @@ export async function executeApiCall(
...
+  try {
+    const response = await callEndpoint(originalEndpoint, payload, credentials, options);
+
+    if (!response.data) {
+      throw new Error("No data returned from API. This could be due to a configuration error.");
+    }
+
+    // In test mode, always evaluate the response
+    if (isTestMode) {
+      const { Documentation } = await import('../../utils/documentation.js');
```
Comment (stefanfaistenauer) [inline packages/core/graphql/resolvers/call.ts]: let's do imports at the top of the page

File: packages/core/graphql/resolvers/call.ts [hunk 4]
```diff
@@ -23,86 +25,173 @@ export async function executeApiCall(
...
 
+  try {
+    const response = await callEndpoint(originalEndpoint, payload, credentials, options);
+
+    if (!response.data) {
+      throw new Error("No data returned from API. This could be due to a configuration error.");
+    }
+
+    // In test mode, always evaluate the response
+    if (isTestMode) {
```
Comment (stefanfaistenauer) [inline packages/core/graphql/resolvers/call.ts]: clarification question: how long does the post processing take? if it takes more than 10ms or so, we should only do that once. right now we do it for the evaluation and then again for the self healing

File: packages/core/graphql/resolvers/call.ts [hunk 5]
```diff
@@ -23,86 +25,173 @@ export async function executeApiCall(
...
+  data: any;
+  endpoint: ApiConfig;
+}> {
   let documentationString = "No documentation provided";
-  if (!integration && isSelfHealing) {
+  if (!integration) {
     logMessage('debug', `Self-healing enabled but no integration provided; skipping documentation-based healing.`, metadata);
-  } else if (integration && integration.documentationPending) {
+  } else if (integration.documentationPending) {
     logMessage('warn', `Documentation for integration ${integration.id} is still being fetched. Proceeding without documentation.`, metadata);
```
Comment (stefanfaistenauer) [inline packages/core/graphql/resolvers/call.ts:36]: in this case, often times when it is being refetched, should we not rather take whatever is there?

File: packages/core/graphql/resolvers/call.ts [hunk 6]
```diff
@@ -23,86 +25,173 @@ export async function executeApiCall(
...
+  const messages: OpenAI.Chat.ChatCompletionMessageParam[] = [
+    {
+      role: "system",
+      content: SELF_HEALING_API_AGENT_PROMPT
+    },
+    {
+      role: "user",
+      content: `Execute this API call successfully. The initial attempt failed with the following error:
+
+ERROR: ${initialError}
```
Comment (stefanfaistenauer) [inline packages/core/graphql/resolvers/call.ts]: let's use xml tags

File: packages/core/llm/prompts.ts [hunk 7]
```diff
@@ -1,3 +1,4 @@
+//LEGACY prompt
 export const PROMPT_MAPPING = `You are an AI that generates JSONata mapping expressions to transform source data structures into target structures.
```
Comment (stefanfaistenauer) [inline packages/core/llm/prompts.ts:2]: do we actually still use that? if not lets cut it

File: packages/core/llm/prompts.ts [hunk 8]
```diff
@@ -130,11 +131,46 @@ Use \n where appropriate for readability.
   - If an object is optional but its fields required, you can add a test and default to {}, but do not set the inner fields to default null.
 Remember: The goal is to create valid JSONata expressions that accurately transform the source data structure into the required target structure. Follow all of these guidelines or I will lose my job.`;
 
-export const API_PROMPT = `You are an API configuration assistant. Generate API details based on instructions and documentation and available variables in a valid JSON format.
+export const MODIFY_STEP_CONFIG_TOOL_PROMPT = `You are an API configuration assistant. Generate API details based on instructions and documentation and available variables in a valid JSON format.
```
Comment (stefanfaistenauer) [inline packages/core/llm/prompts.ts]: same here, do we use the modify step thing?

File: packages/core/llm/prompts.ts [hunk 9]
```diff
@@ -143,120 +179,171 @@ export const API_PROMPT = `You are an API configuration assistant. Generate API
...
+- No authentication: Use authentication: "NONE"
 
-Important: Listen closely to the feedback, identify the cause of the error and adress the cause of the error.
-Make sure to try a fix before generating a new configuration. I will loose my job if I don't get this right.
-</ERROR_HANDLING>
+Most modern APIs use HEADER authentication type with different header formats.
+</AUTHENTICATION>
+
+<RESPONSE_HANDLING>
+- dataPath: The JSON path to extract data from the response (e.g., "data.items", "results", "products.list")
```
Comment (stefanfaistenauer) [inline packages/core/llm/prompts.ts]: do we use data path still?

File: packages/core/llm/prompts.ts [hunk 10]
```diff
@@ -143,120 +179,171 @@ export const API_PROMPT = `You are an API configuration assistant. Generate API
...
 </PAGINATION>
 
 <POSTGRES>
-- You can use the following format to access a postgres database: urlHost: "postgres://<<user>>:<<password>>@<<hostname>>:<<port>>", urlPath: "<<database>>", body: {query: "SELECT...."}
-- For creating the query, use the schema. Consider that some tables need to be joined depending on the instruction.
+For PostgreSQL databases:
+- urlHost: Must use "postgres://" protocol (NOT "https://")
+- urlPath: Database name only
+- method: "POST"
+- body: "{\"query\": \"SELECT * FROM table\"}"
```
Comment (stefanfaistenauer) [inline packages/core/llm/prompts.ts]: please clarify that the query is an example

File: packages/core/llm/prompts.ts [hunk 11]
```diff
@@ -143,120 +179,171 @@ export const API_PROMPT = `You are an API configuration assistant. Generate API
...
+
+// Conditional transformation
+if (sourceData.type === 'batch') {
+  return sourceData.items.map(transformItem);
+} else {
+  return [transformItem(sourceData)];
+}
+\`\`\`
+
+COMMON WORKFLOW TRANSFORMATIONS:
```
Comment (stefanfaistenauer) [inline packages/core/llm/prompts.ts:189]: this is great

File: packages/core/llm/prompts.ts [hunk 12]
```diff
@@ -344,55 +435,145 @@ Important: Avoid using LOOP mode for potentially very large data objects. If you
...
+</DEFENSIVE_JS_TRANSFORMS>
 
-<OUTPUT_FORMAT>
-Important: Your model output must be just the valid JSON without line breaks and tabs, nothing else.
-</OUTPUT_FORMAT>
+<SOAP>
+For SOAP requests:
+- Put the entire XML envelope in the body as a string
+- Include all namespaces and proper XML structure
+- Example body: "<soapenv:Envelope xmlns:soapenv=\"http://schemas.xmlsoap.org/soap/envelope/\">...</soapenv:Envelope>"
```
Comment (stefanfaistenauer) [inline packages/core/llm/prompts.ts:483]: we might want to add an example of how a body code could look like

File: packages/core/utils/api.ts [hunk 13]
```diff
@@ -112,9 +111,11 @@ export async function callEndpoint(
       (Array.isArray(response?.data?.errors) && response?.data?.errors.length > 0)
     ) {
       const error = JSON.stringify(response?.data?.error || response.data?.errors || response?.data || response?.statusText || "undefined");
+      // Mask credentials in the config before logging
+      const maskedConfig = maskCredentials(JSON.stringify(axiosConfig));
```
Comment (stefanfaistenauer) [inline packages/core/utils/api.ts:119]: good

File: packages/core/utils/tools.ts [hunk 14]
```diff
@@ -463,4 +524,44 @@ export function safeHttpMethod(method: any): HttpMethod {
...
+    const script = `
+          const response = JSON.parse(responseJSON);
+          const pageInfo = JSON.parse(pageInfoJSON);
+          const fn = ${stopConditionCode};
+          const result = fn(response, pageInfo);
+          // Ensure we return a boolean
+          result === true;
+      `;
+
+    const shouldStop = await context.eval(script, { timeout: 3000 });
```
Comment (stefanfaistenauer) [inline packages/core/utils/tools.ts]: have we tested this? what is the difference between eval and evalClosure.

File: packages/core/utils/api.ts [hunk 15]
```diff
@@ -45,8 +33,9 @@ export async function callEndpoint(
   let hasMore = true;
   let loopCounter = 0;
   let seenResponseHashes = new Set<string>();
+  const MAX_PAGINATION_REQUESTS = 1000;
```
Comment (stefanfaistenauer) [inline packages/core/utils/api.ts]: move this to defaults.ts

File: packages/core/utils/documentation.ts [hunk 16]
```diff
@@ -93,13 +93,18 @@ export class Documentation {
...
+    chunk_size: number = 0
+  ): string {
     if (documentation.length <= Documentation.MAX_LENGTH) {
       return documentation;
     }
-
-    const CHUNK_SIZE = Documentation.MAX_LENGTH / 10;
-    const MAX_CHUNKS = 10; // Limit number of chunks to return
+    if (chunk_size <= 0) {
+      chunk_size = Documentation.MAX_LENGTH / 10;
```
Comment (stefanfaistenauer) [inline packages/core/utils/documentation.ts]: / 10 is based on the max_chunks being 10 so it should rather be / max_chunks particularly if that is set to 20

File: packages/core/utils/documentation.ts [hunk 17]
```diff
@@ -325,80 +335,98 @@ export class PlaywrightFetchingStrategy implements FetchingStrategy {
...
     // Expanded keywords to catch more relevant documentation pages
     const requiredKeywords = [
       // Auth related
-      "authentication", "authorization", "bearer", "basic", "token",
+      "authentication", "authorization", "bearer", "basic", "token", "api key", "oauth", "private app",
       // Getting started
-      "introduction", "getting started", "quickstart", "guide", "tutorial",
+      "introduction", "getting started", "quickstart", "guide", "guides", "tutorial", "how to", "how-to",
       // API specific
-      "rest", "graphql", "openapi", "endpoints", "reference", "methods", "operations", "objects", "users", "models", "query",
```
Comment (stefanfaistenauer) [inline packages/core/utils/documentation.ts:341]: any reasons we removed these? particularly query and methods (potentially reference for api reference) seem important

File: packages/core/utils/documentation.ts [hunk 18]
```diff
@@ -325,80 +335,98 @@ export class PlaywrightFetchingStrategy implements FetchingStrategy {
...
+      if (rankedLinks.length === 0) break;
+      const nextLink = rankedLinks[0];
 
-      for (const keyword of requiredKeywords) {
-        if (linkTextLower.includes(keyword) || hrefLower.includes(keyword)) {
-          matchCount++;
+      try {
+        const linkResult = await this.fetchPageContentWithPlaywright(nextLink.href, config, metadata);
+        fetchedLinks.add(nextLink.href);
+        if (linkResult?.content) {
```
Comment (stefanfaistenauer) [inline packages/core/utils/documentation.ts]: small nit but maybe we can do a negative pattern here. like if !linkResult.content continue to not have too many brackets

File: packages/core/utils/postgres.ts [hunk 19]
```diff
@@ -27,9 +27,16 @@ function sanitizeDatabaseName(connectionString: string): string {
 
 export async function callPostgres(endpoint: ApiConfig, payload: Record<string, any>, credentials: Record<string, any>, options: RequestOptions): Promise<any> {
   const requestVars = { ...payload, ...credentials };
-  let connectionString = await replaceVariables(composeUrl(endpoint.urlHost, endpoint.urlPath), requestVars);
+  const urlHost = await replaceVariables(endpoint.urlHost, requestVars);
```
Comment (stefanfaistenauer) [inline packages/core/utils/postgres.ts]: not that i am against splitting them up, but out of curiosity - why are we doing it?

File: packages/core/utils/postgres.ts [hunk 20]
```diff
@@ -27,9 +27,16 @@ function sanitizeDatabaseName(connectionString: string): string {
...
+  const urlPath = await replaceVariables(endpoint.urlPath, requestVars);
+  const body = await replaceVariables(endpoint.body, requestVars);
+
+  // Compose and sanitize connection string
+  let connectionString = composeUrl(urlHost, urlPath);
   connectionString = sanitizeDatabaseName(connectionString);
-  const query = await replaceVariables(JSON.parse(endpoint.body).query, requestVars);
+
+  // Parse query from body
+  const query = JSON.parse(body).query;
```
Comment (stefanfaistenauer) [inline packages/core/utils/postgres.ts]: any reason we are not replacing variables in the body? what if query: "SELECT x from z where id=<<userId>>"

File: packages/core/utils/tools.ts [hunk 21]
```diff
@@ -370,6 +370,67 @@ function oldReplaceVariables(template: string, variables: Record<string, any>):
   });
 }
 
+export function validateVariableReferences(
```
Comment (stefanfaistenauer) [inline packages/core/utils/tools.ts]: are we using that function anywhere?

Comment (stefanfaistenauer) [reply packages/core/utils/tools.ts]: i dont quite get what this does lets discuss

File: packages/core/workflow/workflow-builder.ts [hunk 22]
```diff
@@ -71,47 +51,48 @@ export class WorkflowBuilder {
...
-    Credentials available: ${JSON.stringify(Object.entries(int.credentials || {}).reduce((obj, [name, value]) => ({ ...obj, [`${int.id}_${name}`]: value }), {}) || 'None')}
-    ${int.specificInstructions ? `\n    User Instructions for this integration:\n    ${int.specificInstructions}\n` : ''}
-    Documentation:
-    \`\`\`
-    ${processedDoc || 'No documentation content available.'}
-    \`\`\`
-  </${int.id}>`;
+      return baseInfo + `
+  Documentation:
+  \`\`\`
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-builder.ts]: same here lets use xml tags

File: packages/core/workflow/workflow-builder.ts [hunk 23]
```diff
@@ -71,47 +51,48 @@ export class WorkflowBuilder {
...
+  ${processedDoc || 'No documentation content available.'}
+  \`\`\`
+</${int.id}>`;
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-builder.ts:102]: particularly if we use a closing tag here

File: packages/core/workflow/workflow-executor.ts [hunk 24]
```diff
@@ -227,10 +228,19 @@ export class WorkflowExecutor implements Workflow {
       };
 
       if (step.inputMapping) {
-        // Prepare context for JSONata expression
+        // Use JS transform for input mapping
         try {
-          const result = await applyJsonata(mappingContext, step.inputMapping || "$");
-          return result as Record<string, unknown>;
+          const transformResult = await applyTransformationWithValidation(
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-executor.ts]: we might want to rename this function actually if it says applyTransformationWithValidation but actually applies transformation also without validation if you set schema to null

File: packages/core/workflow/workflow-strategies.ts [hunk 25]
```diff
@@ -47,7 +48,7 @@ const directStrategy: ExecutionStrategy = {
         metadata,
         integration
       );
-      const transformedData = await applyJsonata(apiResponse.data, step.responseMapping);
+      const transformedData = await applyJsonata(apiResponse.data, step.responseMapping); //LEGACY: New workflow strategy will not use respone mappings
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-strategies.ts:50]: should we not use applyTransformationWithValidation here?

Comment (stefanfaistenauer) [reply packages/core/workflow/workflow-strategies.ts:50]: to allow response mappings as JS

File: packages/core/workflow/workflow-strategies.ts [hunk 26]
```diff
@@ -85,44 +86,129 @@ const loopStrategy: ExecutionStrategy = {
...
-        loopItems = (await applyTransformationWithValidation(payload, step.loopSelector, null)).data || [];
+        logMessage("error", `No input data found for '${step.id}' - regenerating data selector`, metadata);
+
+        const instruction = `Create a JavaScript function that extracts the array of items to loop over for step: ${step.id}. 
+          
+Step instruction: ${step.apiConfig.instruction}
+
+The function should:
+1. Extract an array of ACTUAL DATA ITEMS (not metadata or property definitions)
+2. Return an empty array if no valid data is found
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-strategies.ts]: lets discuss this

File: packages/core/workflow/workflow-strategies.ts [hunk 27]
```diff
@@ -85,44 +86,129 @@ const loopStrategy: ExecutionStrategy = {
...
+            );
+
+            // Store the successful configuration
+            if (apiResponse.endpoint) {
+              successfulConfig = apiResponse.endpoint;
+              logMessage("debug", `Loop iteration ${i + 1} succeeded, storing config`, metadata);
+            }
+          } else {
+            // We have a successful config, try direct call for efficiency
+            const { callEndpoint } = await import('../utils/api.js');
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-strategies.ts]: lets not use inline imports

File: packages/core/workflow/workflow-strategies.ts [hunk 28]
```diff
@@ -85,44 +86,129 @@ const loopStrategy: ExecutionStrategy = {
...
 
-        const loopPayload = {
+        const loopPayload: Record<string, any> = {
           ...payload,
           currentItem: currentItem,
+          ...flattenObject(currentItem, 'currentItem')
         };
 
         try {
-          const apiResponse = await executeApiCall(
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-strategies.ts:116]: i dont quite get the change here. would executeApiCall not do exactly what we coded here? like, execute once, check if it should evaluate, etc, then self heal if needed. we just have to make sure it does not evaluate all the time though.

File: packages/core/workflow/workflow-tools.ts [hunk 29]
```diff
@@ -0,0 +1,532 @@
...
+                }
+            };
+        }
+
+        const searchResults = Documentation.postProcess(integration.documentation, query);
+
+        const searchTerms = query.toLowerCase().split(/[^a-z0-9]/)
+            .filter(term => term.length >= 3);
+
+        const results = [];
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-tools.ts]: this seems quite odd - does postProcess not handle all of that?

Comment (stefanfaistenauer) [reply packages/core/workflow/workflow-tools.ts]: like it seems like we are just doing the work twice

File: packages/core/workflow/workflow-tools.ts [hunk 30]
```diff
@@ -0,0 +1,532 @@
...
+    logMessage('debug', `submit_tool called - ${apiConfig.method} ${apiConfig.urlHost}${apiConfig.urlPath}`, context);
+
+    // Validate variables before making the API call
+    const availableVariables = [
+        ...Object.keys(credentials || {}),
+        ...Object.keys(payload || {})
+    ];
+
+    const { validateVariableReferences } = await import('../utils/tools.js');
+    const validation = validateVariableReferences(apiConfig, availableVariables);
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-tools.ts]: ahh there we use it. do we need that? lets discuss

File: packages/core/workflow/workflow-tools.ts [hunk 31]
```diff
@@ -0,0 +1,532 @@
...
+
+    logMessage('debug', `submit_tool called - ${apiConfig.method} ${apiConfig.urlHost}${apiConfig.urlPath}`, context);
+
+    // Validate variables before making the API call
+    const availableVariables = [
+        ...Object.keys(credentials || {}),
+        ...Object.keys(payload || {})
+    ];
+
+    const { validateVariableReferences } = await import('../utils/tools.js');
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-tools.ts]: again no inline imports pls

File: packages/core/workflow/workflow-tools.ts [hunk 32]
```diff
@@ -0,0 +1,532 @@
...
+            ...apiConfig,
+            // Preserve the original instruction - don't let self-healing change the step's purpose
+            instruction: originalEndpoint?.instruction || apiConfig.instruction,
+            responseSchema: originalEndpoint?.responseSchema || apiConfig.responseSchema,
+            id: originalEndpoint?.id || apiConfig.id,
+            createdAt: originalEndpoint?.createdAt || new Date(),
+            updatedAt: new Date()
+        };
+
+        const { callEndpoint } = await import('../utils/api.js');
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-tools.ts]: inline import

File: packages/core/workflow/workflow-tools.ts [hunk 33]
```diff
@@ -0,0 +1,532 @@
...
+        const response = await callEndpoint(mergedConfig, payload, credentials, options);
+
+        const finalData = response.data;
+
+        // Always evaluate the response if we have instruction or responseSchema
+        if ((mergedConfig.instruction || mergedConfig.responseSchema) && (options?.testMode || !options?.selfHealing || options.selfHealing === SelfHealingMode.ENABLED || options.selfHealing === SelfHealingMode.REQUEST_ONLY)) {
+            const { evaluateResponse } = await import('../utils/api.js');
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-tools.ts]: again

File: packages/core/workflow/workflow-tools.ts [hunk 34]
```diff
@@ -0,0 +1,532 @@
...
+            updatedAt: new Date()
+        };
+
+        const { callEndpoint } = await import('../utils/api.js');
+        const response = await callEndpoint(mergedConfig, payload, credentials, options);
+
+        const finalData = response.data;
+
+        // Always evaluate the response if we have instruction or responseSchema
+        if ((mergedConfig.instruction || mergedConfig.responseSchema) && (options?.testMode || !options?.selfHealing || options.selfHealing === SelfHealingMode.ENABLED || options.selfHealing === SelfHealingMode.REQUEST_ONLY)) {
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-tools.ts]: lets move the evaluation into a named boolean particularly the second one. what happens if the first (mergedConfig.instruction || mergedConfig.responseSchema) is not true?

File: packages/core/workflow/workflow-builder.ts [hunk 35]
```diff
@@ -121,108 +102,109 @@ ${this.instruction}
...
+          messages
+        };
+
+        // Call the build_workflow tool
+        const result = await executeTool(
+          {
+            id: `build-workflow-${attempts}`,
+            name: 'build_workflow',
+            arguments: {
+              previousError: lastError
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-builder.ts]: this pattern seems of to me. we are getting a previousError. but what could that be? is it important context? something like wrong json? lets discuss.

File: packages/core/workflow/workflow-tools.ts [hunk 36]
```diff
@@ -0,0 +1,532 @@
...
+        const builtWorkflowSchema = zodToJsonSchema(z.object({
+            id: z.string().describe("The workflow ID (e.g., 'stripe-create-order')"),
+            steps: z.array(z.object({
+                id: z.string().describe("Unique camelCase identifier for the step (e.g., 'fetchCustomerDetails')"),
+                integrationId: z.string().describe("The integration ID for this step"),
+                executionMode: z.enum(["DIRECT", "LOOP"]).describe("DIRECT for single execution, LOOP for iterating over collections"),
+                loopSelector: z.string().optional().describe("JavaScript function to select items to loop over. Format: (sourceData) => sourceData.items. Only required if executionMode is LOOP"),
+                inputMapping: z.string().optional().describe("JavaScript function to transform accumulated data before this step. Format: (sourceData) => ({ field1: sourceData.date, field2: sourceData.stepId.data }). Access previous steps via sourceData.stepId. MUST throw descriptive errors if required data is missing - NO defensive programming with null/empty arrays as defaults. If null, all accumulated data passes through."),
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-tools.ts]: do we need this?

File: packages/core/workflow/workflow-tools.ts [hunk 37]
```diff
@@ -0,0 +1,532 @@
...
+                        key: z.string(),
+                        value: z.string()
+                    })).optional().describe("Query parameters as key-value pairs. Use <<variable>> syntax for dynamic values"),
+                    headers: z.array(z.object({
+                        key: z.string(),
+                        value: z.string()
+                    })).optional(),
+                    body: z.string().optional().describe("Format as JSON if not instructed otherwise. Use <<>> to access variables."),
+                    authentication: z.enum(Object.values(AuthType) as [string, ...string[]]).describe("Authentication type: None, Basic, Bearer, OAuth2, or ApiKey"),
+                    dataPath: z.string().optional().describe("JSONPath to extract data from response (e.g., 'data.items' or 'results[*].id')"),
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-tools.ts]: do we still want this?

File: packages/core/workflow/workflow-tools.ts [hunk 38]
```diff
@@ -0,0 +1,532 @@
...
+                        cursorPath: z.string().describe("If cursor_based: The path to the cursor in the response. E.g. cursor.current or next_cursor. If not, set this to \"\""),
+                        stopCondition: z.string().describe("REQUIRED: JavaScript function that determines when to stop pagination. This is the primary control for pagination. Format: (response, pageInfo) => boolean. Return true to STOP pagination. Common patterns: '(response) => response.data.length === 0' (empty page), '(response, pageInfo) => pageInfo.totalFetched >= 100' (limit total), '(response) => !response.has_more' (API flag)")
+                    }).optional()
+                }).describe("Complete API configuration for this step")
+            })).describe("Array of workflow steps with full configuration"),
+            finalTransform: z.string().describe("JavaScript function to transform the final workflow output to match responseSchema. Format: (sourceData) => ({ result: sourceData }). Access step results via sourceData.stepId"),
+            integrationIds: z.array(z.string()).describe("List of all integration IDs used in the workflow"),
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-tools.ts]: would we not have this at the apiconfig level?

Comment (stefanfaistenauer) [reply packages/core/workflow/workflow-tools.ts]: turns out we have it at the step config level so this is probably not required

File: packages/core/workflow/workflow-tools.ts [hunk 39]
```diff
@@ -0,0 +1,532 @@
...
+                    body: {
+                        type: ["string", "object"],
+                        description: "Request body. For REST APIs: Format as JSON string if not instructed otherwise."
+                    },
+                    authentication: {
+                        type: "string",
+                        enum: ["NONE", "HEADER", "QUERY_PARAM", "OAUTH2"],
+                        description: "Authentication type: NONE (no auth), HEADER (for Bearer/Basic/ApiKey in headers), QUERY_PARAM (for API key in URL), or OAUTH2"
+                    },
+                    dataPath: {
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-tools.ts]: outdated no?

File: packages/core/workflow/workflow-tools.ts [hunk 40]
```diff
@@ -0,0 +1,532 @@
...
+                            },
+                            stopCondition: {
+                                type: "string",
+                                description: "JavaScript function that determines when to stop pagination. Format: (response, pageInfo) => boolean"
+                            }
+                        }
+                    },
+                    instruction: {
+                        type: "string",
+                        description: "OPTIONAL for self-healing: Only provide this when creating a new configuration, not when fixing an existing one. The original step instruction will be preserved. If provided, describes what this API call should achieve and what constitutes a successful response."
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-tools.ts]: will it be preserved?

File: packages/core/workflow/workflow-tools.ts [hunk 41]
```diff
@@ -0,0 +1,532 @@
...
+                    headers: {
+                        type: "object",
+                        description: "HTTP headers as key-value pairs. Use <<variable>> syntax for dynamic values",
+                        additionalProperties: { type: "string" }
+                    },
+                    body: {
+                        type: ["string", "object"],
+                        description: "Request body. For REST APIs: Format as JSON string if not instructed otherwise."
+                    },
+                    authentication: {
```
Comment (stefanfaistenauer) [inline packages/core/workflow/workflow-tools.ts]: are we actually using this anywhere?

File: packages/core/eval/utils/setup-manager.ts (hunk 2 above)
Comment (michaelfuest) [reply packages/core/eval/utils/setup-manager.ts]: Yep, required because we crawl way more docs now

File: packages/core/graphql/resolvers/call.ts (hunk 4 above)
Comment (michaelfuest) [reply packages/core/graphql/resolvers/call.ts]: Doesn't take long, but you're right. Changed it to always postProcess once at most in all cases

File: packages/core/llm/prompts.ts (hunk 7 above)
Comment (michaelfuest) [reply packages/core/llm/prompts.ts:2]: We still do in the transform logic, which I have not yet touched (that still has JSONata). Could be something we handle and clean up in this PR, or when we integrate transfroms and extracts into the workflows. Let's discuss

File: packages/core/utils/postgres.ts (hunk 20 above)
Comment (michaelfuest) [reply packages/core/utils/postgres.ts]: we are, but I cleaned this up a little

File: packages/core/utils/tools.ts (hunk 21 above)
Comment (michaelfuest) [reply packages/core/utils/tools.ts]: Added this to catch variable failures deterministically early on, and to provide targeted error feedback and address hallucinated variable usage, but we can remove this and just let the retry loop handle this "organically", let's discuss

File: packages/core/workflow/workflow-executor.ts (hunk 24 above)
Comment (michaelfuest) [reply packages/core/workflow/workflow-executor.ts]: renamed to `transformAndValidateSchema` which is more clear, if the schema is null then there is nothing to validate but tbd

File: packages/core/workflow/workflow-strategies.ts (hunk 25 above)
Comment (michaelfuest) [reply packages/core/workflow/workflow-strategies.ts:50]: Chose not to to make sure that it is clear that response mappings are no longer used. Long-term plan should be to at some point remove them entirely and rely on input mappings and final transforms only imo. May even be a good thing if response mappings in JS fail here, because we don't want to generate them in the first place. TBD

File: packages/core/workflow/workflow-strategies.ts (hunk 26 above)
Comment (michaelfuest) [reply packages/core/workflow/workflow-strategies.ts]: Removed the line, it is definitely problematic. Part of a larger discussion around handling of empty results, specifically in `evaluateResponse` actually. An empty array can be a valid response, but very very rarely is. TBD

File: packages/core/workflow/workflow-strategies.ts (hunk 28 above)
Comment (michaelfuest) [reply packages/core/workflow/workflow-strategies.ts:116]: fixed

File: packages/core/workflow/workflow-tools.ts (hunk 30 above)
Comment (michaelfuest) [reply packages/core/workflow/workflow-tools.ts]: Yup TBD

File: packages/core/workflow/workflow-builder.ts (hunk 35 above)
Comment (michaelfuest) [reply packages/core/workflow/workflow-builder.ts]: Mirrors the existing retry pattern in planning/building without making the whole thing agentic (where you would not need these errors to be manual input. I actually have never seen the LLM fail to come up with a valid workflow, so we could also drop this. TBD

File: packages/core/workflow/workflow-tools.ts (hunk 36 above)
Comment (michaelfuest) [reply packages/core/workflow/workflow-tools.ts]: What are you referring to here? Are you saying that inputMappings and loopSelectors have an overlap, and we could use an input mapping for generating the loop items?

File: packages/core/llm/prompts.ts (hunk 12 above)
Comment (michaelfuest) [reply packages/core/llm/prompts.ts:483]: TBD, would want to have a proper Soap API part of the integration tests

File: packages/core/workflow/workflow-tools.ts (hunk 29 above)
Comment (michaelfuest) [reply packages/core/workflow/workflow-tools.ts]: I disagree, here's why: 

`postProcess` is a post processing of the entire documentation based on context taken from the user instruction, but theoretically returns a tailored summary for LLM context.

The search tool should (and currently is) be used for very targeted queries regarding error messages or endpoints and should also return much smaller results to give back to the tool. The pattern should imo be: Make this tool call fast and return smaller results, and let the LLM decide if the search results contain the info it needs. If not it can call search docs again.

Having said that, let's discuss
